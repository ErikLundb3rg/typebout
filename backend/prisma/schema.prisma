// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Users {
  id           Int            @id @default(autoincrement())
  name         String
  password     String
  createdAt    DateTime       @default(now())
  performances Performances[]
}

model Performances {
  id           Int      @id @default(autoincrement())
  correct      Int
  mistakes     Int
  completed_in Int
  createdAt    DateTime @default(now())
  user         Users    @relation(fields: [userId], references: [id])
  userId       Int
  race         Races    @relation(fields: [raceId], references: [id])
  raceId       Int
}

model Races {
  id           Int            @id @default(autoincrement())
  createdAt    DateTime       @default(now())
  performances Performances[]
  quote        Quotes         @relation(fields: [quotesId], references: [id])
  quotesId     Int
}

model Quotes {
  id         Int          @id @default(autoincrement())
  content    String
  categories Categories[]
  author     Author       @relation(fields: [authorId], references: [id])
  authorId   Int
  Races      Races[]
}

model Author {
  id     Int      @id @default(autoincrement())
  name   String
  Quotes Quotes[]
}

model Categories {
  id     Int      @id @default(autoincrement())
  name   String
  quotes Quotes[]
}
