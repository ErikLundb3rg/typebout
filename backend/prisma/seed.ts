import { PrismaClient } from '@prisma/client'
import quoteFile from '../quotes/quotes.json'

const prisma = new PrismaClient()

const main = async () => {
  // delete all quotes

  quoteFile.quotes.forEach(async (quote) => {
    await prisma.quotes.upsert({
      where: { content: quote.content },
      update: {},
      create: {
        content: quote.content,
        author: {
          connectOrCreate: {
            where: {
              name: quote.author
            },
            create: {
              name: quote.author
            }
          }
        }
      }
    })
  })
}

main()
  .then(async () => {
    await prisma.$disconnect()
  })
  .catch(async (error) => {
    console.error(error)
    await prisma.$disconnect()
    process.exit(1)
  })
