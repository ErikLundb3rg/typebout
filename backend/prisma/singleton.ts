import { PrismaClient } from '@prisma/client';
import { mockDeep, mockReset, DeepMockProxy } from 'jest-mock-extended';

import db from './client';

jest.mock('./client', () => ({
  __esModule: true,

  default: mockDeep<PrismaClient>()
}));

export const prismaMock = db as unknown as DeepMockProxy<PrismaClient>;

beforeEach(() => {
  mockReset(prismaMock);
});
