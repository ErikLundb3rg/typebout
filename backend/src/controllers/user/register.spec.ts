/* eslint-disable @typescript-eslint/no-explicit-any */
import { prismaMock } from '../../../prisma/singleton'
import { register } from './index'
import bcrypt from 'bcrypt'
import {
  generateAccessToken,
  generateRefreshToken
} from '../../auth/util/verifyers'
import axios from 'axios'

jest.mock('axios')

describe('Register Controller', () => {
  let req: any = {}
  let res: any = {}
  let user: any = {}

  beforeEach(() => {
    req = {}
    res = {}
    user = {
      id: 1,
      username: 'username',
      password: 'password',
      createdAt: new Date('2023')
    }
    ;(generateAccessToken as jest.Mock) = jest
      .fn()
      .mockReturnValue('accesstoken')
    ;(generateRefreshToken as jest.Mock) = jest.fn()
    res.cookie = jest.fn()
    ;(axios.post as jest.Mock).mockResolvedValue({ data: { success: true } })
  })

  test('is successful with correct input', async () => {
    req.body = {
      username: 'username',
      password: 'password',
      confirmPassword: 'password',
      captcha: 'captcha'
    }
    prismaMock.users.findUnique.mockResolvedValue(null)
    prismaMock.users.create.mockResolvedValue(user)
    ;(bcrypt.hash as jest.Mock) = jest.fn().mockResolvedValue('hashed_password')

    expect(await register(req, res)).toMatchSnapshot()
  })

  test('is unsuccessful with empty input', async () => {
    req.body = {
      username: '',
      password: '',
      confirmPassword: '',
      captcha: 'captcha'
    }

    expect(await register(req, res)).toMatchSnapshot()
  })

  test('is unsuccessful with confirmedPassword not matching password', async () => {
    req.body = {
      username: 'username',
      password: 'password',
      confirmPassword: 'passwordDifferent',
      captcha: 'captcha'
    }

    expect(await register(req, res)).toMatchSnapshot()
  })

  test('unsuccessful with already existing user', async () => {
    req.body = {
      username: 'username',
      password: 'password',
      confirmPassword: 'password',
      captcha: 'captcha'
    }
    prismaMock.users.findUnique.mockResolvedValue(user)

    expect(await register(req, res)).toMatchSnapshot()
  })
})
