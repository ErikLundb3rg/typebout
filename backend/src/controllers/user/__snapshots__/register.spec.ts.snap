// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Controller is successful with correct input 1`] = `
{
  "data": {
    "accessToken": "accesstoken",
    "user": {
      "createdAt": 2023-01-01T00:00:00.000Z,
      "id": 1,
      "password": "password",
      "username": "username",
    },
  },
  "message": "User signup was successful",
  "ok": true,
  "status": 200,
}
`;

exports[`Register Controller is unsuccessful with confirmedPassword not matching password 1`] = `
{
  "data": {
    "errors": {
      "captcha": "",
      "confirmPassword": "Confirm password does not match password",
      "password": "",
      "username": "",
    },
  },
  "message": null,
  "ok": false,
  "status": 400,
}
`;

exports[`Register Controller is unsuccessful with empty input 1`] = `
{
  "data": {
    "errors": {
      "captcha": "",
      "confirmPassword": "",
      "password": "",
      "username": ""username" is not allowed to be empty",
    },
  },
  "message": null,
  "ok": false,
  "status": 400,
}
`;

exports[`Register Controller unsuccessful with already existing user 1`] = `
{
  "data": {
    "errors": {
      "username": "A user with this username already exists",
    },
  },
  "message": null,
  "ok": false,
  "status": 400,
}
`;
